### ISSUE TITLE
Improve rendering of cabal parser errors in cabal-install

### ISSUE THREAD

**Issue by hvr** [haskell/cabal#5572]
Currently, when `cabal` encounters a syntax error in package descriptions, it merely dumps the Haskell data structure of the error, e.g.

```
$ cabal new-build
CabalFileParseError "./U.cabal" [PError (Position 4 1) "\"the input\" (line 4, column 1):\nunexpected \"}\"\nexpecting \":\", section parameter, \"{\", field or section name or end of file"] (Just (mkVersion [2,4])) []
```

instead, it'd be much nicer if we could instead mimic something like GHC's current parse error message style (IMO, it'd be generally a good idea to have a consistent error style across GHC and Cabal because when you use cabal you get confrontend w/ lots of error messages from GHC and then it's more desirable to *not* have completely different styles presented... :-) ):

```
U.hs:4:1: error:
    Parse error: module header, import declaration
    or top-level declaration expected.
  |
4 | asd
  | ^^^
```

cc @ChShersh @vrom911 as I believe this to be your area of interest

**Comment by ChShersh** [comment]
@m-renaud Feel free to do so! I have this issue open in tab in my browser for 2 months but didn't have time to look at it closely :disappointed:

**Comment by ulysses4ever** [comment]
@m-renaud do you plan to improve on this even more by any chance?

**Comment by ulysses4ever** [comment]
It's not clear to me how to get a raw parser error these days: whatever garbage I put I get a civil warning `Ignoring section: "blah"`. So, maybe (just maybe!) this one could be closed.

**Comment by gbaz** [comment]
If I recall, this class of errors from parsing section arguments is still really bad? https://github.com/haskell/cabal/issues/7846

**Comment by ulysses4ever** [comment]
Thanks for the pointer! #7846 starts with:

```
Warning: MissingH.cabal:1:1:
unexpected SecArgName (Position 204 7) "base"
Errors encountered when parsing cabal file ./MissingH.cabal:

MissingH.cabal:1:1: error:
unexpected SecArgName (Position 204 7) "base"

    1 | cabal-version: 1.12
      | ^
Error: cabal: parse error
```

which is a pretty decent parse error rendering I'd say (even if a bit excessive) and, surely a great improvement over the example that started this ticket:

```
CabalFileParseError "./U.cabal" [PError (Position 4 1) "\"the input\" (line 4, column 1):\nunexpected \"}\"\nexpecting \":\", section parameter, \"{\", field or section name or end of file"] (Just (mkVersion [2,4])) []
```

It seems to me that this ticket can be closed and more specific tickets (like #7846 asking for proper propagation of error locations in the messages) could be used for further improvements.

**Comment by Mikolaj** [comment]
Let me close in favour of https://github.com/haskell/cabal/issues/7846.
